torch/lib/THD/base/channels/DataChannelTCP.cpp

```diff
@@ -196,70 +196,99 @@ DataChannelTCP::~DataChannelTCP()
 
 
 void DataChannelTCP::listen(std::uint16_t port = 0) {
-  SYSCHECK(_socket = ::socket(PF_INET, SOCK_STREAM, 0))
+  struct addrinfo hints, *res = NULL;
+
+  memset(&hints, 0x00, sizeof(hints));
+  hints.ai_flags = AI_PASSIVE;
+  hints.ai_family = AF_UNSPEC; // either IPv4 or IPv6
+  hints.ai_socktype = SOCK_STREAM; // TCP
+
+  // `getaddrinfo` will sort addresses according to RFC 3484 and can be tweeked
```

Comment:
Of `getaddrinfo`? It's quite long and consists of numbers of rules: https://github.com/lattera/glibc/blob/master/sysdeps/posix/getaddrinfo.c#L1591

I decided to remove IPv6 preference over IPv4 and just stick to some RFC which I guess is thought out. So instead greedy IPv6 addresses picking we get more carefully picked addresses.

And as it is pointed out in comment, when someone wants to change the order, it just change order in some specified file `/etc/gai.conf` instead of changing code (which is not that easy for a normal user)